src/core/dag.rs

```diff
@@ -736,15 +737,16 @@ impl DAG {
             *fun_parents.as_ptr() = DLL::singleton(ParentCell::Left(app));
             *arg_parents.as_ptr() = DLL::singleton(ParentCell::Right(app));
           }
-          let fun = go(*fun, ctx.clone(), fun_parents);
-          let arg = go(*arg, ctx, arg_parents);
+          let fun = go((*fun).clone(), ctx.clone(), fun_parents);
```

Comment:
These clones are unnecessary, I think. If you do:

```
let (fun, arg) = *terms;
```

to get the Box contents, then you can just do

```
let fun = go(fun, ctx.clone(), fun_parents);
```